driver-core/src/main/com/mongodb/internal/connection/InternalStreamConnection.java

```diff
@@ -678,7 +712,11 @@ private void updateSessionContext(final SessionContext sessionContext, final Res
         }
     }
 
-    private MongoException translateWriteException(final Throwable e) {
+    private MongoException translateWriteException(final Throwable e, final OperationContext operationContext) {
+        if (e instanceof MongoSocketWriteTimeoutException && operationContext.getTimeoutContext().hasExpired()) {
```

Comment:
I initially missed this, but this comment https://github.com/mongodb/mongo-java-driver/pull/1278#discussion_r1426087259 applies here as well.